### ISSUE TITLE
Incorrect filter not detected

### ISSUE THREAD

**Issue by friedl** [rsyslog/rsyslog#1046]
When having incorrect filter operations, the configuration check doesn't produce a warning/error. Also, the debug output does not show this. 

Config example:
`if $msg contains 'bla1' or 'bla2' then {
        action(type="omfile" file="/var/log/bla12.log")
} else {
        if $msg contains 'bla3' or 'bla4' then {
                action(type="omfile" file="/var/log/bla34.log")
        } else {
                if $msg contains 'bla5' then {
                        action(type="omfile" file="/var/log/bla5.log")
                } else {
                        if $msg contains 'bla6' then {
                                action(type="omfile" file="/var/log/bla6.log")
                        } else {
                                action(type="omfile" file="/var/log/blarest.log")
                                }
                        }
                }
        }
`

When reading this, it seems logical. The problem is, that after the "or" operator a complete operation is expected with property, comparison and value. Instead, just a value is provided, which results in a boolean comparison that always evaluates to true. Thus, all operations further down the configuration will not be carried out, even if the values from the message did not trigger the first rule. 

To have this working correctly, the syntax for the filters with multiple values must be different. It must be either of these:
`if $msg contains ["bla1", "bla2"] then {}

if $msg contains 'bla1' or $msg contains 'bla2' then {}`

**Comment by davidelang** [comment]
On Thu, 19 May 2016, friedl wrote:

> `if $msg contains 'bla1' or 'bla2' then {
<snip>
> When reading this, it seems logical. The problem is, that after the "or" operator a complete operation is expected with property, comparison and value. Instead, just a value is provided, which results in a boolean comparison that always evaluates to true. Thus, all operations further down the configuration will not be carried out, even if the values from the message did not trigger the first rule.
>
> To have this working correctly, the syntax for the filters with multiple values must be different. It must be either of these:
> `if $msg contains ["bla1", "bla2"] then {}
>
> if $msg contains 'bla1' or $msg contains 'bla2' then {}`

The problem is that

   if $msg contains 'bla1' or $!variable then {

where $!variable is set from an earlier test rather than duplicating the test in 
many places.

is a very valid approach to take.

It's very common for scripting languages to accept a variable or bare value in a 
test (if 'True' then), admittedly only commonly used for testing purposes, but 
available all the same.

So I'm not sure that this really should be detected as a syntax error. It's 
commonly a logic error, but it's not possible to be sure it's not intended.

David Lang

**Comment by friedl** [comment]
> It's very common for scripting languages to accept a variable or bare value in a test (if 'True' then), admittedly only commonly used for testing purposes, but available all the same.

Yes, you are right. But, for the goal I was trying to achieve, this would lead to a "misbehaviour", which (in my opinion) needs some kind of change to at least point it out.

**Comment by rgerhards** [comment]
I think the key point to consider here is that @friedl's sample does include a (text) **constant**. In essence, it boils down to ```if expression or literal```, which is ```if expression or false``` which is ```if expression```. In that context, I think at least a warning would be good to give (gcc and clang do the same in this case).

**Comment by davidelang** [comment]
I think a warning on a constant, bare variable, or math expression (as opposed 
to function or comparison) is a fair thing to do.

David Lang
